feat: adjust theme types to match internal and regular usage (#3279)
callstack/react-native-paper: docs/pages/2.theming.md

## TypeScript

> By default TypeScript works well whenever you change the value of built-in theme's properties. It gets more complicated when you want to extend the theme's properties or change their types. In order to fully support TypeScript, you will need to follow the guide that fits your use-case most accurately:

Reviewer: ```suggestion
By default, TypeScript works well whenever you change the value of the built-in theme's properties. It gets more complicated when you want to extend the theme's properties or change their types. In order to fully support TypeScript, you will need to follow the guide that fits your use-case most accurately:
```

> There are two supported ways of overriding theme:

Reviewer: ```suggestion
There are two supported ways of overriding the theme:
```

1. **Simple built-in theme overrides** - when you only customize the values and the whole theme schema remains the same
2. **Advanced theme overrides** - when you _add new properties_ or _change the built-in schema shape_

### Simple built-in theme overrides

> You can provide a `theme` prop with a theme object with same properties as the default theme:

Reviewer: ```suggestion
You can provide a `theme` prop with a theme object with the same properties as the default theme:
```

```js
import * as React from 'react';
import { MD3LightTheme, Provider as PaperProvider } from 'react-native-paper';
import App from './src/App';

const theme = {
  ...MD3LightTheme, // or MD3DarkTheme
  roundness: 2,
  colors: {
    ...MD3LightTheme.colors,
    primary: '#3498db',
    secondary: '#f1c40f',
    tertiary: '#a1b2c3',
  },
};

export default function Main() {
  return (
    <PaperProvider theme={theme}>
      <App />
    </PaperProvider>
  );
}
```

### Advanced theme overrides

> If you need to modify the built-in theme schema by adding a new property or changing it type, you need to follow these steps:

Reviewer: ```suggestion
If you need to modify the built-in theme schema by adding a new property or changing its type, you need to follow these steps:
```

1. Pass your theme overrides to the Provider component

```ts
import * as React from 'react';
import { MD3LightTheme, Provider as PaperProvider } from 'react-native-paper';
import App from './src/App';

const theme = {
  ...MD3LightTheme,

  // Specify a custom property
  custom: 'property',

  // Specify a custom property in nested object
  colors: {
    ...MD3LightTheme.colors,
    brandPrimary: '#fefefe',
    brandSecondary: 'red',
  },
};

export default function Main() {
  return (
    <PaperProvider theme={theme}>
      <App />
    </PaperProvider>
  );
}
```

2. Create a typed `useAppTheme()` hook in your project

```ts
import * as React from 'react';

> import { MD3LightTheme, Provider as PaperProvider } from 'react-native-paper';

Reviewer: ```suggestion
import { MD3LightTheme, Provider as PaperProvider, useTheme } from 'react-native-paper';
```

}
```

> 3. Use the `useAppTheme()` hook in your components

Reviewer: ```suggestion
3. Start using the `useAppTheme()` hook across your components in the whole app
```

or something similar, just want to **highlight** that then `useAppTheme` should be the one and only hook for getting the theme

```ts
import * as React from 'react';
import { useAppTheme } from './App';

export default function HomeScreen() {
  const {
    colors: { brandPrimary },
  } = useAppTheme();

  return <View style={{ backgroundColor: brandPrimary }}>...</View>;
}
```

## Material Design 2

Using Material Design 2 is **fully supported in React Native Paper v5+**.

### Simple setup

> In order to use Material Design 2 theme you can just pass `{ version: 2 }` to the PaperProvider theme prop:

Reviewer: ```suggestion
In order to use the Material Design 2 theme you can just pass `{ version: 2 }` to the PaperProvider theme prop:
```

Specifying `{ version: 2 }` tells React Native Paper to use the built in Material Design 2 theme, so you don't have to fully extend it on your own.

### Advanced setup

> As with any theme, you can also specify your custom properties within Material Design 2 theme:

Reviewer: ```suggestion
As with any theme, you can also specify your custom properties within the Material Design 2 theme:
```

```js
import * as React from 'react';
import { MD2LightTheme, Provider as PaperProvider } from 'react-native-paper';
import App from './src/App';

export default function Main() {
  const theme = {
    ...MD2LightTheme,

    // Specify a custom property
    custom: 'property',

    // Specify a custom nested property
    colors: {
      ...MD2LightTheme.colors,
      primary: '#fefefe',
    },
  };

  return (
    <PaperProvider theme={theme}>
      <App />
    </PaperProvider>
  );
}
```

### Typescript

Due to the amount of changes in the theme's schema shape it falls into the [Advanced theme overrides](#advanced-theme-overrides) category. The steps are identical as with any advanced theme, just make sure to extend the built-in `MD2LightTheme` or `MD2DarkTheme` instead of `MD3LightTheme` or `MD3DarkTheme`.

The final example for Material Design 2 would look like this:

```ts
import * as React from 'react';

> import { MD2LightTheme, Provider as PaperProvider } from 'react-native-paper';

Reviewer: ```suggestion
import { MD2LightTheme, Provider as PaperProvider, useTheme } from 'react-native-paper';
```
